pkgs/build-support/dotnet/make-nuget-deps/default.nix

```diff
@@ -1,14 +1,34 @@
-{ linkFarmFromDrvs, fetchurl }:
+{ linkFarmFromDrvs, fetchurl, stdenvNoCC, zip }:
 { name, nugetDeps ? import sourceFile, sourceFile ? null }:
 linkFarmFromDrvs "${name}-nuget-deps" (nugetDeps {
   fetchNuGet = { pname, version, sha256 ? "", hash ? ""
     , url ? "https://www.nuget.org/api/v2/package/${pname}/${version}" }:
-    fetchurl {
+    stdenvNoCC.mkDerivation {
       name = "${pname}.${version}.nupkg";
-      # There is no need to verify whether both sha256 and hash are
-      # valid here, because nuget-to-nix does not generate a deps.nix
-      # containing both.
-      inherit url sha256 hash;
+      src = fetchurl {
+        # There is no need to verify whether both sha256 and hash are
+        # valid here, because nuget-to-nix does not generate a deps.nix
+        # containing both.
+        inherit url sha256 hash;
+      };
+      dontUnpack = true;
+      nativeBuildInputs = [ zip ];
+      installPhase = "cp $src $out";
+      # NuGet.org edits packages by signing them during upload, which makes
+      # those packages nondeterministic. Strip out the signature file.
+      # Signing logic is https://github.com/NuGet/NuGet.Client/blob/128a5066b1438627ac69a2ffe9de564b2c09ee4d/src/NuGet.Core/NuGet.Packaging/Signing/Archive/SignedPackageArchiveIOUtility.cs#L518
+      # Non-NuGet.org sources might not have a signature file.
+      fixupPhase = ''
+        runHook preFixup
+        if zip --show-files "$out" | grep --quiet --line-regexp '[[:space:]]\+\.signature\.p7s'; then
+          local tmp_file
+          tmp_file=$(mktemp tmp-XXXXXXXX.zip)
+          cp "$out" "$tmp_file"
+          zip -d "$tmp_file" .signature.p7s
```

Comment:
I just couldn't make `zip -d` work on a path that's in the Nix store. I tried in about eight different configurations (including with `install -Dm644` in the install phase). So we're back to using a temp file.

Reply:
```
diff --git a/pkgs/build-support/dotnet/make-nuget-deps/default.nix b/pkgs/build-support/dotnet/make-nuget-deps/default.nix
index a9661b4274c8..3a739b72265b 100644
--- a/pkgs/build-support/dotnet/make-nuget-deps/default.nix
+++ b/pkgs/build-support/dotnet/make-nuget-deps/default.nix
@@ -13,7 +13,11 @@ linkFarmFromDrvs "${name}-nuget-deps" (nugetDeps {
       };
       dontUnpack = true;
       nativeBuildInputs = [ zip ];
-      installPhase = "cp $src $out";
+      installPhase = ''
+        runHook preInstall
+        install -Dm644 "$src" "$out"
+        runHook postInstall
+      '';
       # NuGet.org edits packages by signing them during upload, which makes
       # those packages nondeterministic depending on which source you
       # get them from. We fix this by stripping out the signature file.
@@ -22,11 +26,7 @@ linkFarmFromDrvs "${name}-nuget-deps" (nugetDeps {
       fixupPhase = ''
         runHook preFixup
         if zip --show-files "$out" | grep --quiet --line-regexp '[[:space:]]\+\.signature\.p7s'; then
-          local tmp_file
-          tmp_file=$(mktemp tmp-XXXXXXXX.zip)
-          cp "$out" "$tmp_file"
-          zip -d "$tmp_file" .signature.p7s
-          mv "$tmp_file" "$out"
+          zip -d "$out" .signature.p7s
         fi
         runHook postFixup
       '';
```

this seems to work for me.  is it something darwin specific that's failing?

Reply:
I guess we should assume that `zip -d` might want to write a temp file next to the target zip, which would be bad in the nix store.

[...] I actually just tested this with strace and it's doing 

```
rename("/nix/store/ziMkPS9K", "/nix/store/qk8kffricm52gj6agd0lii7j3g74i95f-Microsoft.TestPlatform.ObjectModel.17.8.0.nupkg") = 0
```

I'm not sure why that's allowed in the linux sandbox to be honest.

I like to avoid using `$TMPDIR` in derivations if at all possible, because it ends up polluting the user's directory when run in `nix-shell`.

I think what I'd suggest in this case is to copy the package into the build dir in `unpackPhase`, and then strip the signature in `patchPhase`?

Reply:
Yeah, with that patch:

```
error: builder for '/nix/store/i5pp5v40s5mcp5p7vw6w9cahdnn4cb0r-Microsoft.AspNetCore.Connections.Abstractions.2.2.0.nupkg.drv' failed with exit code 10;
       last 10 log lines:
       > no configure script, doing nothing
       > Running phase: buildPhase
       > no Makefile or custom buildPhase, doing nothing
       > Running phase: installPhase
       > Running phase: fixupPhase
       > calling 'preFixup' function hook '_moveToShare'
       > calling 'preFixup' function hook '_multioutDocs'
       > calling 'preFixup' function hook '_multioutDevs'
       > zip I/O error: Permission denied
```

Reply:
>I guess we should assume that zip -d might want to write a temp file next to the target zip, which would be bad in the nix store.

Would it make sense to use zip’s `-b`/`--temp-path` flag?
https://linux.die.net/man/1/zip#:~:text=%2Db%20path%0A%2D%2Dtemp%2Dpath%20path,destination%20file%20system%2E